From 65b6b241dcac900e996171897b46fecb6551cd27 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Mon, 20 Oct 2025 11:35:51 +0100
Subject: [PATCH] vddk: Don't use FNM_PATHNAME when matching export parameter

Commit 18866d92d1 ("vddk: Add export parameter") added recently allows
a single VDDK plugin instance to serve multiple files on the VMware
server.  The NBD client chooses which file to serve by sending an NBD
export name.

As a mild protection against an attacker who has access to the NBD
endpoint from requesting any file from the server (and not just those
related to a single virtual machine), you are able to specify an
export wildcard.  The NBD export name from the client is compared to
the wildcard before allowing the connection to read anything.

The wildcard was matched (before this commit) using:

  fnmatch (export_wildcard, nbd_export_name, FNM_PATHNAME)

As a concrete example:

 - Virtual machine with two disks called:
   "[esx] esx-win11/esx-win11.vmdk"
   "[esx] esx-win11/esx-win11-2.vmdk"

 - nbdkit vddk export="\[esx\] esx-win11/*.vmdk"

NBD client requests "[esx] esx-win11/esx-win11.vmdk" => allowed
(because this export name matches the wildcard).

NBD client requests "[esx] unrelated guest/disk.vmdk" => denied
(because this export name does not match the wildcard).

Note the effect of FNM_PATHNAME, as described in the man page:

  FNM_PATHNAME
    If this flag is set, match a slash in string only with a slash in
    pattern and not by  an  asterisk  (*)  or  a  question  mark  (?)
    metacharacter,  nor  by  a  bracket  expression ([]) containing a
    slash.

Now consider a different guest where the disks are located at
different subdirectory depths:

 - Virtual machine with two disks called:
   "[esx] esx-win10/esx-win10.vmdk"
   "[esx] esx-win10/disk2/esx-win10-2.vmdk"

 - nbdkit vddk export=???

It is not possible to construct an export wildcard that matches both
disk names, because the special character '/' in the export name must
match exactly one '/' in the wildcard, but the two paths have
different numbers of '/' characters.

The best solution I can come up with for this is to relax the
restriction by removing the FNM_PATHNAME parameter of fnmatch.  Thus
(after this commit is applied):

 - Virtual machine with two disks called:
   "[esx] esx-win10/esx-win10.vmdk"
   "[esx] esx-win10/disk2/esx-win10-2.vmdk"

 - nbdkit vddk export="\[esx\] esx-win10/*.vmdk"

As part of this change, I also tested how "//" and "/../" in paths
behaves, so that we weren't opening up a hole where someone could
request "[esx] esx-win10/../unrelated guest/disk.vmdk".  It turns out
that VMware does not parse these as paths.  It seems to treat these
parameters only as simple strings, and any variation in the string is
rejected.

Note also that the export wildcard is really a backstop in case
regular permissions on the NBD endpoint fail.  It's not supposed to
offer complete security.

Reported-by: Ming Xie
Related: https://issues.redhat.com/browse/RHEL-121728
Related: https://github.com/libguestfs/virt-v2v/commit/ba86b22c75a65240fdb65ac2e91c472c0f68f78e
(cherry picked from commit 01b429c412504a491a4b5cc009a9c2e906f993ef)
---
 plugins/vddk/vddk.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/plugins/vddk/vddk.c b/plugins/vddk/vddk.c
index e16813e4..6c667dc7 100644
--- a/plugins/vddk/vddk.c
+++ b/plugins/vddk/vddk.c
@@ -639,6 +639,7 @@ vddk_dump_plugin (void)
 
   printf ("vddk_default_libdir=%s\n", VDDK_LIBDIR);
   printf ("vddk_has_nfchostport=1\n");
+  printf ("vddk_has_fnm_pathname=0\n");
 
   /* Because load_library (false) we might not have loaded VDDK, in
    * which case we didn't set library_version.  Note this cannot
@@ -744,7 +745,7 @@ vddk_open (int readonly)
       goto err0;
     }
 
-    r = fnmatch (export_wildcard, nbd_export_name, FNM_PATHNAME);
+    r = fnmatch (export_wildcard, nbd_export_name, 0);
     switch (r) {
     case 0: /* OK */ break;
     case FNM_NOMATCH:
-- 
2.47.3

